File: src/easy/relative_sort_array.zig
const std = @import("std");

const expectEqualSlices = std.testing.expectEqualSlices;

const Allocator = std.mem.Allocator;

pub fn relativeSortArray(comptime T: type, allocator: Allocator, arr1: []const T, arr2: []const T) ?[]const T {
    var freq = allocator.alloc(T, 1_001) catch return null;
    var result = allocator.alloc(T, arr1.len) catch return null;
    var index: usize = 0;
    defer allocator.free(freq);
    @memset(freq, 0);
    for (arr1) |num| {
        const position: usize = @intCast(num);
        freq[position] += 1;
    }
    for (arr2) |num| {
        const position: usize = @intCast(num);
        var count = freq[position];
        while (count > 0) : (count -= 1) {
            freq[position] -= 1;
            result[index] = num;
            index += 1;
        }
    }
    for (freq, 0..) |num, i| {
        const position: T = @intCast(i);
        var count = num;
        while (count > 0) : (count -= 1) {
            result[index] = position;
            index += 1;
        }
    }
    return result;
}
test "relatively sort [2, 3, 1, 3, 2, 4, 6, 7, 9, 2, 19]" {
    const arr1 = [_]i32{ 2, 3, 1, 3, 2, 4, 6, 7, 9, 2, 19 };
    const arr2 = [_]i32{ 2, 1, 4, 3, 9, 6 };
    const expected = [_]i32{ 2, 2, 2, 1, 4, 3, 3, 9, 6, 7, 19 };
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [28, 6, 22, 8, 44, 17]" {
    const arr1 = [_]i32{ 28, 6, 22, 8, 44, 17 };
    const arr2 = [_]i32{ 22, 28, 8, 6 };
    const expected = [_]i32{ 22, 28, 8, 6, 17, 44 };
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [5, 3, 8, 1, 9, 2]" {
    const arr1 = [_]i32{ 5, 3, 8, 1, 9, 2 };
    const arr2 = [_]i32{};
    const expected = [_]i32{ 1, 2, 3, 5, 8, 9 };
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [5, 5, 2, 2, 8, 8]" {
    const arr1 = [_]i32{ 5, 5, 2, 2, 8, 8 };
    const arr2 = [_]i32{ 2, 5, 8 };
    const expected = [_]i32{ 2, 2, 5, 5, 8, 8 };
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [42]" {
    const arr1 = [_]i32{42};
    const arr2 = [_]i32{42};
    const expected = [_]i32{42};
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [100, 1, 100, 2]" {
    const arr1 = [_]i32{ 100, 1, 100, 2 };
    const arr2 = [_]i32{ 1, 2 };
    const expected = [_]i32{ 1, 2, 100, 100 };
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [3, 3, 3, 3, 3]" {
    const arr1 = [_]i32{ 3, 3, 3, 3, 3 };
    const arr2 = [_]i32{3};
    const expected = [_]i32{ 3, 3, 3, 3, 3 };
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [0]" {
    const arr1 = [_]i32{0};
    const arr2 = [_]i32{0};
    const expected = [_]i32{0};
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort []" {
    const arr1 = [_]usize{};
    const arr2 = [_]usize{ 1, 2, 3 };
    const expected = [_]usize{};
    const allocator = std.testing.allocator;
    const result = relativeSortArray(usize, allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(usize, &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [7, 1, 7, 1, 7]" {
    const arr1 = [_]i32{ 7, 1, 7, 1, 7 };
    const arr2 = [_]i32{ 7, 1 };
    const expected = [_]i32{ 7, 7, 7, 1, 1 };
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [999, 0, 500, 1000, 0]" {
    const arr1 = [_]i32{ 999, 0, 500, 1000, 0 };
    const arr2 = [_]i32{ 0, 1000 };
    const expected = [_]i32{ 0, 0, 1000, 500, 999 };
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}
test "relatively sort [1, 2, 3, 4, 5]" {
    const arr1 = [_]i32{ 1, 2, 3, 4, 5 };
    const arr2 = [_]i32{ 5, 4, 3, 2, 1 };
    const expected = [_]i32{ 5, 4, 3, 2, 1 };
    const allocator = std.testing.allocator;
    const result = relativeSortArray(@TypeOf(arr1[0]), allocator, &arr1, &arr2);
    if (result) |actual| {
        defer allocator.free(actual);
        try expectEqualSlices(@TypeOf(arr1[0]), &expected, actual);
    } else return error.SkipZigTest;
}

File: src/easy/relative_sort_array.rs
use std::collections::HashMap;

pub fn relative_sort_array(arr1: Vec<i32>, arr2: Vec<i32>) -> Vec<i32> {
    let mut result = vec![0; arr1.len()];
    let mut freq: HashMap<i32, i32> = HashMap::with_capacity(arr1.len());
    let mut index: usize = 0;
    for &num in &arr1 {
        *freq.entry(num).or_insert(0) += 1;
    }
    for &num in &arr2 {
        if let Some(&count) = freq.get(&num) {
            for _ in 0..count {
                result[index] = num;
                index += 1;
            }
            freq.remove(&num);
        }
    }
    let mut keys: Vec<i32> = freq.keys().copied().collect();
    keys.sort();

    for num in keys {
        if let Some(&count) = freq.get(&num) {
            for _ in 0..count {
                result[index] = num;
                index += 1;
            }
        }
    }
    result
}
#[cfg(test)]
mod tests {
    use super::relative_sort_array;
    use rstest::rstest;
    #[rstest]
    #[case(&[2, 3, 1, 3, 2, 4, 6, 7, 9, 2, 19], &[2, 1, 4, 3, 9, 6], &[2, 2, 2, 1, 4, 3, 3, 9, 6, 7, 19])]
    #[case(&[28, 6, 22, 8, 44, 17], &[22, 28, 8, 6], &[22, 28, 8, 6, 17, 44])]
    #[case(&[5, 3, 8, 1, 9, 2], &[], &[1, 2,3, 5, 8, 9])]
    #[case(&[5, 5, 2, 2, 8, 8], &[2, 5, 8], &[2, 2, 5, 5, 8, 8])]
    #[case(&[42], &[42], &[42])]
    #[case(&[100, 1, 100, 2], &[1, 2], &[1, 2, 100, 100])]
    #[case(&[3, 3, 3, 3, 3], &[3], &[3, 3, 3, 3, 3])]
    #[case(&[0], &[0], &[0])]
    #[case(&[], &[1, 2, 3], &[])]
    #[case(&[7, 1, 7, 1, 7], &[7, 1], &[7, 7, 7, 1, 1])]
    #[case(&[999, 0, 500, 1000, 0], &[0, 1000], &[0, 0, 1000, 500, 999])]
    #[case(&[1, 2, 3, 4, 5], &[5, 4, 3, 2, 1], &[5, 4, 3, 2, 1])]

    fn test_relative_sort_array(
        #[case] arr1: &[i32],
        #[case] arr2: &[i32],
        #[case] expected: &[i32],
    ) {
        assert_eq!(
            relative_sort_array(arr1.to_vec(), arr2.to_vec()),
            expected.to_vec()
        );
    }
}
